pkg/controller/job/jobcontroller.go

```diff
@@ -431,7 +431,41 @@ func (jm *JobController) syncJob(key string) error {
 	}
 
 	var manageJobErr error
-	if pastActiveDeadline(&job) {
+
+	jobTimeout := pastActiveDeadline(&job)
+
+	if jobNeedsSync && job.DeletionTimestamp == nil && !jobTimeout {
+		active, manageJobErr = jm.manageJob(activePods, succeeded, &job)
+	}
+	completions := succeeded
+	jobComplete := false
+	if job.Spec.Completions == nil {
+		// This type of job is complete when any pod exits with success.
+		// Each pod is capable of
+		// determining whether or not the entire Job is done.  Subsequent pods are
+		// not expected to fail, but if they do, the failure is ignored.  Once any
+		// pod succeeds, the controller waits for remaining pods to finish, and
+		// then the job is complete.
+		if succeeded > 0 && active == 0 {
+			jobComplete = true
+		}
+	} else {
+		if completions >= *job.Spec.Completions {
+			jobComplete = true
+			if active > 0 {
+				jm.recorder.Event(&job, v1.EventTypeWarning, "TooManyActivePods", "Too many active pods running after completion count reached")
+			}
+			if completions > *job.Spec.Completions {
+				jm.recorder.Event(&job, v1.EventTypeWarning, "TooManySucceededPods", "Too many succeeded pods running after completion count reached")
+			}
+		}
+	}
+
+	if jobComplete {
+		job.Status.Conditions = append(job.Status.Conditions, newCondition(batch.JobComplete, "", ""))
+		now := metav1.Now()
+		job.Status.CompletionTime = &now
+	} else if jobTimeout {
```

Comment:
With this new logic, a job can be both complete and timeout, and such job won't kill its active pods while it should